matlab_simulation/code v1.0/ME597-6-EKFSLAM/ekf_slam.m

```diff
@@ -113,102 +116,94 @@
+                count=count+1;
+                % rearrange the order of estimates so that those of the newly
+                % detected feture is placed on top, right after the vehicle
+                % states
+                if count==1
+                   mu=[mu(1:3);
+                        mu(1)+y(2*(i-1)+1,t)*cos(y(2*i,t)+mu(3));
+                        mu(2)+y(2*(i-1)+1,t)*sin(y(2*i,t)+mu(3))];
+                else
+                    mu=[mu(1:3);
+                        mu(1)+y(2*(i-1)+1,t)*cos(y(2*i,t)+mu(3));
+                        mu(2)+y(2*(i-1)+1,t)*sin(y(2*i,t)+mu(3));
+                        mu(4:end)];  % augment measured features into states
+                end
+                % rearrange the covariance matrix. Those of the newly
+                % discovered features are set to be S0mm, a predefined
+                % value. Off-diagonal terms are set to 0.
+                temp=zeros(length(S)+2,length(S)+2);
+                temp(1:3,1:3)=S(1:3,1:3);
+                temp(4:5,4:5)=S0mm;
+                if length(S)>3
+                    temp(6:end,1:3)=S(4:end,1:3);
+                    temp(1:3,6:end)=S(1:3,4:end);
+                    temp(6:end,6:end)=S(4:end,4:end);
+                end
+                S=temp;
                 newfeature(i) = 0;
+                % rearrange map, newfeature, y, and flist so that they
+                % follow the same order as mu
+                [map, newfeature, y, flist]=rearrangeMap(map,newfeature,y,flist,i,count);
+                % after rearranging, this newly discovered feature is the
+                % first feature in the list
+                j=1;
+                disp([num2str(count) ' of the ' num2str(M) ' features have been detected'])
             end
+           
             % Linearization
             % Predicted range
-            dx = mu(3+2*(i-1)+1)-mu(1);
-            dy = mu(3+2*i)-mu(2);
+            dx = mu(3+2*(j-1)+1)-mu(1);
+            dy = mu(3+2*j)-mu(2);
             rp = sqrt((dx)^2+(dy)^2);
-
-            Fi = zeros(5,N);
-            Fi(1:n,1:n) = eye(n);
-            Fi(4:5,3+2*(i-1)+1:3+2*i) = eye(2);
-            Ht = [ -dx/rp ...
-                -dy/rp ...
-                0 ...
-                dx/rp ...
-                dy/rp;
-                dy/rp^2 ...
-                -dx/rp^2 ...
-                -1 ...
-                -dy/rp^2 ...
-                dx/rp^2]*Fi;
-
-            I = y(2*(i-1)+1:2*i,t)-[rp;
-                (atan2(dy,dx) - mu(3))];
-            
+            Ht = range_bearing_meas_linearized_model(mu,j);
+            I = y(2*(j-1)+1:2*j,t)- range_bearing_meas_model(mu(1:3),mu((3+2*(j-1)+1):3+2*j));
  
             % Measurement update
             K = S*Ht'*inv(Ht*S*Ht'+Qi);
             mu = mu + K*I;
-            S = (eye(n+2*M)-K*Ht)*S;
+            S = (eye(length(S))-K*Ht)*S;
+            
+            % In cases if S bemoes not positive definite, manually make it
```

Comment:
minor edit, suggest 'is not positive definite' over 'becomes not positive definite'